dbService.go

```diff
@@ -0,0 +1,136 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+type dbService interface {
+	DumpCollectionTo(database, collection string, writer io.Writer) error
+	RestoreCollectionFrom(database, collection string, reader io.Reader) error
+}
+
+type mongoService struct {
+	connectionString string
+	mgoLib           mongoLib
+	bsonService      bsonService
+	mongoTimeout     time.Duration
+	rateLimit        time.Duration
+	batchLimit       int
+}
+
+func newMongoService(connectionString string, mgoLib mongoLib, bsonService bsonService, mongoTimeout time.Duration, rateLimit time.Duration, batchLimit int) *mongoService {
+	return &mongoService{
+		connectionString: connectionString,
+		mgoLib:           mgoLib,
+		bsonService:      bsonService,
+		mongoTimeout:     mongoTimeout,
+		rateLimit:        rateLimit,
+		batchLimit:       batchLimit,
+	}
+}
+
+func (m *mongoService) DumpCollectionTo(database, collection string, writer io.Writer) error {
+	session, err := m.mgoLib.DialWithTimeout(m.connectionString, m.mongoTimeout)
+	if err != nil {
+		return err
+	}
+
+	defer session.Close()
+
+	start := time.Now()
+	log.Infof("backing up %s/%s", database, collection)
+
+	iter := session.SnapshotIter(database, collection, nil)
+	for {
```

Comment:
Shouldn't we check if `iter` contains an error before trying to retrieve an element?

Reply:
I think we should but this will impact the loop too much, we tested this like it is, so i'd rather treat this in an upcoming PR.

Reply:
I've done it before the loop, ok.